### ISSUE TITLE
[BUG] spring-data-cosmos wont allow null partition Id on initial save but will allow on update

### ISSUE THREAD

**Issue by Blackbaud-EricSlater** [Azure/azure-sdk-for-java#15291]
**Describe the bug**
If I have an entity with a partition key defined, and the value of that partition key is null when I save the entity for the first time then I get an BadRequest exception saying "PartitionKey extracted from document doesn't match the one specified in the header".

If i set a value on the partition key and save it, then set the partition key to null and save it again. It saves successfully. 

Upon inspection of the code, it looks like the save path for entities that are new (dont have an id or no version field) go down a code path that sets the partition key header to UNDEFINED instead of NULL. ON update of an existing entity, a different code path is taken that properly set the header value to NULL

***Exception or Stack Trace***
com.azure.cosmos.implementation.BadRequestException: ["PartitionKey extracted from document doesn't match the one specified in the header"], {"userAgent":"azsdk-java-cosmos/4.3.2-beta.2 MacOSX/10.15.6 JRE/1.8.0_162","requestLatencyInMs":595,"requestStartTimeUTC":"16 Sep 2020 20:03:22.760","requestEndTimeUTC":"16 Sep 2020 20:03:23.355","connectionMode":"DIRECT","responseStatisticsList":[{"storeResult":{"storePhysicalAddress":"rntbd://cdb-ms-prod-southcentralus1-fd3.documents.azure.com:14410/apps/1865814a-b42a-4543-8e5f-5f5d8188e242/services/fc4c8a56-d1cd-4e23-ad88-d6ef812bff0e/partitions/a8ca4c3a-d7bb-4be5-ae16-d0c3206f20da/replicas/132447534086115570p/","lsn":1,"globalCommittedLsn":1,"partitionKeyRangeId":"0","isValid":true,"statusCode":400,"subStatusCode":1001,"isGone":false,"isNotFound":false,"isInvalidPartition":false,"requestCharge":1.24,"itemLSN":-1,"sessionToken":"-1#1","exception":"[\"PartitionKey extracted from document doesn't match the one specified in the header\"]","transportRequestTimeline":null},"requestResponseTimeUTC":"16 Sep 2020 20:03:23.355","requestResourceType":"Document","requestOperationType":"Create"}],gatewayStatistics":null,"systemInformation":{"usedMemory":"143270 KB","availableMemory":"3585114 KB","processCpuLoad":"40.46003016591252 %","systemCpuLoad":"NaN %"}}
		at com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdRequestManager.messageReceived(RntbdRequestManager.java:742)
		at com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdRequestManager.channelRead(RntbdRequestManager.java:174)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
		at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
		at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
		at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
		at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1526)
		at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1275)
		at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1322)
		at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:501)
		at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:440)
		at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
		at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
		at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
		at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
		at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
		at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		... 1 common frames omitted

**To Reproduce**

***Code Snippet***
This is a test that demonstrates the issue. It is written in Groovy using the spock framework, but the code is easy enough to read and understand even if you cant run it....

```
def "reproduce null partition key issue"() {
	when:
	ExamplePartitionedEntity examplePartitionedEntity = aRandom.examplePartitionedEntity()
					.environmentId(null)
					.build()

	and:
	examplePartitionedRepository.save(examplePartitionedEntity)

	then:
	thrown(Exception)

	when:
	examplePartitionedEntity.environmentId = aRandom.environmentId()

	and:
	examplePartitionedRepository.save(examplePartitionedEntity)

	then:
	noExceptionThrown()

	when:
	examplePartitionedEntity.environmentId = null

	and:
	examplePartitionedRepository.save(examplePartitionedEntity)

	then:
	noExceptionThrown()
}
```

Where ExamplePartitionedEntity is...

```
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = PARTITIONED_TEST_COLLECTION_NAME)
public class ExamplePartitionedEntity implements Persistable<String> {

    @Id
    @GeneratedValue
    private String id;
    @PartitionKey
    private String environmentId;
    private Integer randomField;

    @Override
    public boolean isNew() {
        return id == null;
    }
}

```
 and the repository is....

```
@Repository
public interface ExamplePartitionedRepository extends CosmosRepository<ExamplePartitionedEntity, String> {

```

**Expected behavior**
A null partition key should be a valid value on the first save of an object.

**Setup (please complete the following information):**
 - OS: iOS, linux, windows
 - IDE : not important but used  intellij
 - 3.0.0-beta.1

**Comment by joshfree** [comment]
Thanks for the very detailed bug report @Blackbaud-EricSlater.  @kushagraThapar from the `Cosmos` team will follow up.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater , we just released (yesterday) v3.0.0-beta.2 version of azure-spring-data-cosmos library. I modified some new entity creation code path there. Can you please test it with latest version ?
Also, in the latest version, there is a change in artifact id. The new maven coordinates are here: https://mvnrepository.com/artifact/com.azure/azure-spring-data-cosmos/3.0.0-beta.2

**Comment by kushagraThapar** [comment]
Adding @mbhaskar  to this thread. @mbhaskar  can you please help @Blackbaud-EricSlater here ?

**Comment by Blackbaud-EricSlater** [comment]
Re-ran the test and confirmed that this is fixed in 3.0.0-beta.2. Thanks for the heads up on the new release. I will close the issue.
